internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -976,3 +996,81 @@ func (r *PersistentVolumeClaimReconciler) getScheduleFromPVC(
 
 	return ""
 }
+
+// checkAnnotationForValue checks if the given object has the specified annotation
+// with the expected value.
+func checkAnnotationForValue(obj metav1.Object, key, expected string) bool {
+	if val, ok := obj.GetAnnotations()[key]; ok && val == expected {
+		return true
+	}
+	return false
+}
+
+// hasValidStorageClassName checks if the provided PersistentVolumeClaim has a non-empty StorageClassName.
+func hasValidStorageClassName(pvc *corev1.PersistentVolumeClaim) bool {
+	return pvc.Spec.StorageClassName != nil && len(*pvc.Spec.StorageClassName) > 0
+}
+
+// checkDisabledByAnnotation checks if the given object has an annotation
+// that disables the functionality represented by the provided annotationKey.
+func (r *PersistentVolumeClaimReconciler) checkDisabledByAnnotation(
+	ctx context.Context,
+	logger *logr.Logger,
+	pvc *corev1.PersistentVolumeClaim,
+	annotationKey string) (bool, error) {
+
+	if r.SchedulePrecedence == util.ScheduleSCOnly {
+		if !hasValidStorageClassName(pvc) {
+			return false, nil
+		}
+		storageClassName := *pvc.Spec.StorageClassName
+
+		sc := &storagev1.StorageClass{}
+		err := r.Client.Get(ctx, client.ObjectKey{Name: storageClassName}, sc)
+		if err != nil {
+			logger.Error(err, "Failed to get StorageClass", "StorageClass", storageClassName)
+			return false, err
+		}
+
+		if checkAnnotationForValue(sc, annotationKey, "false") {
+			return true, nil
+		}
+
+		return false, nil
+	}
+
+	// Else, we follow the regular precedence
+	// Check on PVC
+	if checkAnnotationForValue(pvc, annotationKey, "false") {
+		return true, nil
+	}
+
+	// Check on Namespace
+	ns := &corev1.Namespace{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: pvc.Namespace}, ns)
+	if err != nil {
+		logger.Error(err, "Failed to get Namespace", "Namespace", pvc.Namespace)
+		return false, err
+	}
+	if checkAnnotationForValue(ns, annotationKey, "false") {
+		return true, nil
+	}
+
+	// Check on SC
```

Comment:
Avoid the duplicate, the check on SC has been done above.

Reply:
That check is evaluated only when precedence is "sc-only". We still need a check on SC for the other case, no?

Reply:
I meant the same piece of code is duplicated again. you could do something like
```
disabledAtStorageClass = validate()
if SCOnly {
    return disabledAtStorageClass
}

// existing flow
// check at pvc
// check at namespace

return disabledAtStorageClass
```

something like this?

Reply:
Ah, right. Done. The updated code fetches SC only once.